reference/preconditioner/batch_jacobi_kernels.hpp.inc

```diff
@@ -0,0 +1,166 @@
+    const size_type k, const matrix::Csr<ValueType, int>* const first_sys_csr,
+    const int* const cumulative_block_storage, const int* const block_pointers,
+    int* const blocks_pattern)
+{
+    const int* row_ptrs = first_sys_csr->get_const_row_ptrs();
+    const int* col_idxs = first_sys_csr->get_const_col_idxs();
+
+    const int row_idx_st = block_pointers[k];
+    const int row_idx_end = block_pointers[k + 1];  // exclusive
+
+    const auto offset =
+        detail::batch_jacobi::get_block_offset(k, cumulative_block_storage);
+    const auto stride = detail::batch_jacobi::get_stride(k, block_pointers);
+
+    // pattern is stored in row-major order
+    for (int row_idx = row_idx_st; row_idx < row_idx_end; row_idx++) {
+        for (int i = row_ptrs[row_idx]; i < row_ptrs[row_idx + 1]; i++) {
+            const auto col_idx = col_idxs[i];
+
+            if (col_idx >= row_idx_st && col_idx < row_idx_end) {
+                // (row_idx, col_idx) is the part of the diagonal block
+                // storing the position/index of this element at place: (row_idx
+                // - row_idx_st, col_idx - row_idx_st)
+                const auto dense_block_row = row_idx - row_idx_st;
+                const auto dense_block_col = col_idx - row_idx_st;
+
+                blocks_pattern[offset + dense_block_row * stride +
+                               dense_block_col] = i;
+            }
+        }
+    }
+}
+
+
+template <typename ValueType>
+inline void swap_rows(const int size, std::vector<ValueType>& mat,
+                      const int row1, const int row2)
+{
+    for (int c = 0; c < size; c++) {
+        std::swap(mat[row1 * size + c], mat[row2 * size + c]);
+    }
+}
+
+
+template <typename ValueType>
+inline int choose_pivot(const int k, const int block_size,
+                        const std::vector<ValueType>& temp_dense_block)
+{
+    int piv = k;
+    ValueType piv_val = temp_dense_block[k * block_size + k];
+    for (int r = k + 1; r < block_size; r++)  // uncoalesced accesses
+    {
+        ValueType curr_val = temp_dense_block[r * block_size + k];
+        if (abs(piv_val) < abs(curr_val)) {
+            piv = r;
+            piv_val = curr_val;
+        }
+    }
+    return piv;
+}
+
+
+template <typename ValueType>
+inline void invert_dense_block(const int block_size,
+                               std::vector<ValueType>& temp_dense_block,
+                               std::vector<int>& col_perm)
+{
+    // Gauss-Jordan elimination with pivoting
+    // the block is stored in the row-major order, with row stride = block_size
+    for (int k = 0; k < block_size; k++) {
+        const int piv = choose_pivot(k, block_size, temp_dense_block);
+        if (k != piv) {
+            swap_rows(block_size, temp_dense_block, k, piv);
+            std::swap(col_perm[piv], col_perm[k]);
+        }
+        const ValueType d = temp_dense_block[k * block_size + k];
+
+        // SCAL (scaling the kth col)
```

Comment:
what are these SCAL/GER comments? Are those references to a paper or something? If so, then also put it here otherwise, remove the SCAL, since that one is clear, and replace GER with what is actually done there (I guess dense matrix * matrix)

Reply:
those are BLAS nomenclature. SCAL = scaling, GER = general rank 1 update, but will update the docstring to be more clear